File: backend/StudioManagement.Data/Entities/Quotation.cs
namespace StudioManagement.Data.Entities;

public class Quotation : ITenantEntity
{
    public int QuotationId { get; set; }
    public int StudioId { get; set; }
    public string QuotationNumber { get; set; } = null!;
    public int CustomerId { get; set; }
    public int? EventId { get; set; }
    public DateTime QuotationDate { get; set; }
    public DateTime? ValidUntil { get; set; }
    public decimal Subtotal { get; set; }
    public decimal Discount { get; set; }
    public decimal TaxAmount { get; set; }
    public decimal GrandTotal { get; set; }
    public string Status { get; set; } = null!;
    public string? TermsAndConditions { get; set; }

    // QuotationPriceDisplays: how this quotation's PDF shows prices (this quotation only).
    public string PriceDisplay { get; set; } = "Detailed";

    // A total typed in by hand for a "Total only" quotation. When set it IS the quotation's grand
    // total (approved amount, balances, reports); the item prices stay stored for reference.
    public decimal? ManualTotal { get; set; }
    public DateTime CreatedAt { get; set; }
    public DateTime UpdatedAt { get; set; }
    public byte[] RowVersion { get; set; } = null!;

    public Studio Studio { get; set; } = null!;
    public Customer Customer { get; set; } = null!;
    public Event? Event { get; set; }
    public ICollection<QuotationItem> Items { get; set; } = new List<QuotationItem>();
}

File: frontend/src/types/quotation.ts
export type PriceDisplay = "Detailed" | "TotalOnly";

export const QUOTATION_STATUSES = ["Draft", "Sent", "Accepted", "Rejected", "Expired", "Cancelled"] as const;

export type QuotationStatus = (typeof QUOTATION_STATUSES)[number];

export interface QuotationItem {
  quotationItemId: number;
  serviceId: number | null;
  // The catalog service's name, or the typed name of a custom line.
  serviceName: string;
  isCustom: boolean;
  quantity: number;
  unitPrice: number;
  total: number;
  notes: string | null;
}

export interface Quotation {
  quotationId: number;
  quotationNumber: string;
  customerId: number;
  customerName: string;
  customerMobileNumber: string;
  eventId: number | null;
  eventVenue: string | null;
  quotationDate: string;
  validUntil: string | null;
  subtotal: number;
  discount: number;
  taxAmount: number;
  grandTotal: number;
  status: QuotationStatus;
  termsAndConditions: string | null;
  // How this quotation's PDF shows prices (this quotation only).
  priceDisplay: PriceDisplay;
  // Typed total of a Total-only quotation; when set it is the grand total.
  manualTotal: number | null;
  items: QuotationItem[];
  createdAt: string;
  updatedAt: string;
}

export interface QuotationItemRequest {
  // A catalog service, or leave it out and give customName for a custom line.
  serviceId?: number;
  customName?: string;
  quantity: number;
  unitPrice: number;
  notes?: string;
}

export interface CreateQuotationRequest {
  customerId: number;
  eventId?: number;
  quotationDate: string;
  validUntil?: string;
  discount: number;
  taxAmount: number;
  status: QuotationStatus;
  termsAndConditions?: string;
  priceDisplay?: PriceDisplay;
  manualTotal?: number;
  items: QuotationItemRequest[];
}

export type UpdateQuotationRequest = CreateQuotationRequest;
